pkg/apis/apps/validation/validation.go

```diff
@@ -153,12 +153,27 @@ func ValidateStatefulSetUpdate(statefulSet, oldStatefulSet *apps.StatefulSet) fi
 	restoreStrategy := statefulSet.Spec.UpdateStrategy
 	statefulSet.Spec.UpdateStrategy = oldStatefulSet.Spec.UpdateStrategy
 
+	restoreVolumeClaimTemplates := make([]api.PersistentVolumeClaim, len(statefulSet.Spec.VolumeClaimTemplates))
```

Comment:
All of these changes should be feature gated. Potentially either by introducing a new gate or using existing volume expansion feature gate.